File: FrontEnd/src/utils/descriptionMatching.ts
import type { TransactionDescriptionStat } from '../services/transactionDescriptionsService'

export interface ApproximateMatch {
  descriptionStart: string
  matchingTransactionCount: number
}

// A candidate description is an approximate match of `description` if it starts with the same
// characters as `description`, up to and including a space - i.e. the match only counts at a
// whole-word boundary, never mid-word (so "COLES 0717..." vs "COLES 0760..." matches on "COLES",
// not on the coincidentally-shared "07" digits). Where more than one boundary is supported,
// the longest (most precise) one wins - e.g. two "DAVID JONES ..." descriptions match each other
// on "DAVID JONES", which takes precedence over a weaker "DAVID"-only match against an unrelated
// "DAVID CAMERON ..." description.
//
// `otherDescriptions` is the deduplicated per-description stats list (one entry per unique
// description string, each carrying how many real transactions have it) - not a per-transaction
// list. The entry for `description` itself is still a valid candidate: if two+ transactions
// share the exact same description (the common case for real bank data), that's a single stat
// entry with `transactionCount` > 1, and it should still trigger a match against the other
// transaction(s) it represents - it only fails to qualify when `transactionCount` is 1, meaning
// this is the only transaction with that description.
export function findApproximateMatch(description: string, otherDescriptions: TransactionDescriptionStat[]): ApproximateMatch | null {
  const boundaries: number[] = []
  for (let i = 0; i < description.length; i++) {
    if (description[i] === ' ') {
      boundaries.push(i + 1)
    }
  }
  boundaries.sort((a, b) => b - a)

  for (const boundary of boundaries) {
    const prefix = description.slice(0, boundary)
    let matchingTransactionCount = 0

    for (const stat of otherDescriptions) {
      if (!stat.description.startsWith(prefix)) {
        continue
      }
      // The stat for `description` itself also represents the transaction being edited - only
      // the transactions *other* than that one count as a match.
      matchingTransactionCount += stat.description === description ? stat.transactionCount - 1 : stat.transactionCount
    }

    if (matchingTransactionCount > 0) {
      return { descriptionStart: prefix.trimEnd(), matchingTransactionCount }
    }
  }

  return null
}

File: iosApp/iosApp/Utils/DescriptionMatching.swift
import Foundation

// Direct port of FrontEnd/src/utils/descriptionMatching.ts. Keep the two in sync.

struct ApproximateMatch: Equatable {
    let descriptionStart: String
    let matchingTransactionCount: Int
}

enum DescriptionMatching {
    // A candidate description approximately matches `description` if it starts with the same
    // characters up to and including a space - a whole-word boundary, never mid-word. The longest
    // (most precise) boundary that yields any match wins. `otherDescriptions` is the deduplicated
    // per-description stats list; the entry for `description` itself still counts, minus the one
    // transaction being edited.
    static func findApproximateMatch(
        description: String,
        otherDescriptions: [TransactionDescriptionStat]
    ) -> ApproximateMatch? {
        let chars = Array(description)
        var boundaries: [Int] = []
        for i in chars.indices where chars[i] == " " {
            boundaries.append(i + 1)
        }
        boundaries.sort(by: >)

        for boundary in boundaries {
            let prefix = String(chars[0..<boundary])
            var matchingTransactionCount = 0

            for stat in otherDescriptions where stat.description.hasPrefix(prefix) {
                // The stat for `description` itself also represents the transaction being edited -
                // only the *other* transactions it covers count.
                matchingTransactionCount += stat.description == description
                    ? stat.transactionCount - 1
                    : stat.transactionCount
            }

            if matchingTransactionCount > 0 {
                var trimmed = prefix
                while trimmed.hasSuffix(" ") { trimmed.removeLast() }
                return ApproximateMatch(descriptionStart: trimmed, matchingTransactionCount: matchingTransactionCount)
            }
        }

        // No word-boundary match - only possible when `description` has no spaces. A single-token
        // description can still be a genuine exact duplicate of another transaction's identical
        // description (exact match only, not a prefix - "Netflix" must not match "Netflix Extra").
        if let exact = otherDescriptions.first(where: { $0.description == description }), exact.transactionCount > 1 {
            return ApproximateMatch(descriptionStart: description, matchingTransactionCount: exact.transactionCount - 1)
        }

        return nil
    }
}
